Prevent non-test targets from depending on test targets

### Is it reproducible with SwiftPM command-line tools: `swift build`, `swift test`, `swift package` etc?

- [x] Confirmed reproduction steps with SwiftPM CLI. The description text _must_ include reproduction steps with either of command-line SwiftPM commands, `swift build`, `swift test`, `swift package` etc.

### Description

Currently, a developer can have a non-test target depend on an test target, and Swift PM will happily try to compile, and fail



### Expected behavior

An appropriate error message stating a target cannot depend on a test target

### Actual behavior

```
❯ swift build --package-path Fixtures/Miscellaneous/DependencyTargetDependsOnTestTarget
[1/1] Planning build
Building for debugging...
error: emit-module command failed with exit code 1 (use -v to see invocation)
/Users/bkhouri/Documents/git/public/swiftlang/swift-package-manager/Fixtures/Miscellaneous/DependencyTargetDependsOnTestTarget/Sources/DependencyTargetDependsOnTestTarget/DependencyTargetDependsOnTestTarget.swift:4:8: error: compiling for macOS 10.13, but module 'DependencyTargetDependsOnTestTargetTests' has a minimum deployment target of macOS 14.0: /Users/bkhouri/Documents/git/public/swiftlang/swift-package-manager/Fixtures/Miscellaneous/DependencyTargetDependsOnTestTarget/.build/arm64-apple-macosx/debug/Modules/DependencyTargetDependsOnTestTargetTests.swiftmodule
2 | // https://docs.swift.org/swift-book
3 |
4 | import DependencyTargetDependsOnTestTargetTests
  |        `- error: compiling for macOS 10.13, but module 'DependencyTargetDependsOnTestTargetTests' has a minimum deployment target of macOS 14.0: /Users/bkhouri/Documents/git/public/swiftlang/swift-package-manager/Fixtures/Miscellaneous/DependencyTargetDependsOnTestTarget/.build/arm64-apple-macosx/debug/Modules/DependencyTargetDependsOnTestTargetTests.swiftmodule
5 |
/Users/bkhouri/Documents/git/public/swiftlang/swift-package-manager/Fixtures/Miscellaneous/DependencyTargetDependsOnTestTarget/Sources/DependencyTargetDependsOnTestTarget/DependencyTargetDependsOnTestTarget.swift:4:8: error: compiling for macOS 10.13, but module 'DependencyTargetDependsOnTestTargetTests' has a minimum deployment target of macOS 14.0: /Users/bkhouri/Documents/git/public/swiftlang/swift-package-manager/Fixtures/Miscellaneous/DependencyTargetDependsOnTestTarget/.build/arm64-apple-macosx/debug/Modules/DependencyTargetDependsOnTestTargetTests.swiftmodule
2 | // https://docs.swift.org/swift-book
3 |
4 | import DependencyTargetDependsOnTestTargetTests
  |        `- error: compiling for macOS 10.13, but module 'DependencyTargetDependsOnTestTargetTests' has a minimum deployment target of macOS 14.0: /Users/bkhouri/Documents/git/public/swiftlang/swift-package-manager/Fixtures/Miscellaneous/DependencyTargetDependsOnTestTarget/.build/arm64-apple-macosx/debug/Modules/DependencyTargetDependsOnTestTargetTests.swiftmodule
5 |
```

### Steps to reproduce

1. extract the attached [repro.tgz](https://github.com/user-attachments/files/19692504/repro.tgz)
2. `swift build --package-path <path/to/extracted/repo>`

### Swift Package Manager version/commit hash

_No response_

### Swift & OS version (output of `swift --version ; uname -a`)
```
❯ uname -a                                                                                       
Darwin BKs-Work-M1Pro-MacBook-Pro-2.local 24.4.0 Darwin Kernel Version 24.4.0: Tue Mar 11 22:26:20 PDT 2025; root:xnu_development-11417.101.12~4/DEVELOPMENT_ARM64_T6000 arm64

❯ swift --version           
Apple Swift version 6.1 (swift-6.1-RELEASE)
Target: arm64-apple-macosx15.0
```

--- a/Sources/PackageLoading/ManifestLoader+Validation.swift
+++ b/Sources/PackageLoading/ManifestLoader+Validation.swift
@@ -309,9 +309,9 @@ extension Basics.Diagnostic {
         return .error("\(messagePrefix) in dependencies of target '\(targetName)'; valid packages are: \(validPackages.map{ "\($0.descriptionForValidation)" }.joined(separator: ", "))")
     }
 
-    static func invalidDependencyOnTestTarget(dependency: String, targetName: String) -> Self {
+    static func invalidDependencyOnTestTarget(dependency: Module.Dependency, targetName: String) -> Self {
         .error(
-            "Invalid dependency: '\(targetName)' cannot depend on test target dependency '\(dependency)'. Only test targets can depend on other test targets"
+            "Invalid dependency: '\(targetName)' cannot depend on test target dependency '\(dependency.name)'. Only test targets can depend on other test targets"
         )
     }
 

--- a/Sources/PackageLoading/ManifestLoader+Validation.swift
+++ b/Sources/PackageLoading/ManifestLoader+Validation.swift
@@ -309,6 +309,10 @@ extension Basics.Diagnostic {
         return .error("\(messagePrefix) in dependencies of target '\(targetName)'; valid packages are: \(validPackages.map{ "\($0.descriptionForValidation)" }.joined(separator: ", "))")
     }
 
+    static func invalidDependencyOnTestTarget(dependency: String, targetName: String) -> Self {
+        return .error("Invalid dependency: '\(targetName)' cannot depend on test target dependency '\(dependency)'. Only test targets can depend on other test targets")
+    }
+
     static func invalidBinaryLocation(targetName: String) -> Self {
         .error("invalid location for binary target '\(targetName)'")
     }

--- a/Sources/PackageLoading/ManifestLoader+Validation.swift
+++ b/Sources/PackageLoading/ManifestLoader+Validation.swift
@@ -310,7 +310,9 @@ extension Basics.Diagnostic {
     }
 
     static func invalidDependencyOnTestTarget(dependency: String, targetName: String) -> Self {
-        return .error("Invalid dependency: '\(targetName)' cannot depend on test target dependency '\(dependency)'. Only test targets can depend on other test targets")
+        .error(
+            "Invalid dependency: '\(targetName)' cannot depend on test target dependency '\(dependency)'. Only test targets can depend on other test targets"
+        )
     }
 
     static func invalidBinaryLocation(targetName: String) -> Self {

--- a/Sources/PackageLoading/PackageBuilder.swift
+++ b/Sources/PackageLoading/PackageBuilder.swift
@@ -948,17 +948,20 @@ public final class PackageBuilder {
                         ))
                 }
         }
-        
+
         // Ensure non-test targets do not depend on test targets.
         // Only test targets are allowed to have dependencies on other test targets.
         if !potentialModule.isTest {
             for dependency in dependencies {
                 if let depTarget = dependency.module, depTarget.type == .test {
-                    self.observabilityScope.emit(.invalidDependencyOnTestTarget(dependency: dependency.name, targetName: potentialModule.name))
+                    self.observabilityScope.emit(.invalidDependencyOnTestTarget(
+                        dependency: dependency.name,
+                        targetName: potentialModule.name
+                    ))
                 }
             }
         }
-        
+
         // Create the build setting assignment table for this target.
         let buildSettings = try self.buildSettings(
             for: manifestTarget,

--- a/Sources/PackageLoading/PackageBuilder.swift
+++ b/Sources/PackageLoading/PackageBuilder.swift
@@ -948,7 +948,17 @@ public final class PackageBuilder {
                         ))
                 }
         }
-
+        
+        // Ensure non-test targets do not depend on test targets.
+        // Only test targets are allowed to have dependencies on other test targets.
+        if !potentialModule.isTest {
+            for dependency in dependencies {
+                if let depTarget = dependency.module, depTarget.type == .test {
+                    self.observabilityScope.emit(.invalidDependencyOnTestTarget(dependency: dependency.name, targetName: potentialModule.name))
+                }
+            }
+        }
+        
         // Create the build setting assignment table for this target.
         let buildSettings = try self.buildSettings(
             for: manifestTarget,

--- a/Sources/PackageLoading/PackageBuilder.swift
+++ b/Sources/PackageLoading/PackageBuilder.swift
@@ -955,7 +955,7 @@ public final class PackageBuilder {
             for dependency in dependencies {
                 if let depTarget = dependency.module, depTarget.type == .test {
                     self.observabilityScope.emit(.invalidDependencyOnTestTarget(
-                        dependency: dependency.name,
+                        dependency: dependency,
                         targetName: potentialModule.name
                     ))
                 }

--- a/Tests/PackageGraphTests/ModulesGraphTests.swift
+++ b/Tests/PackageGraphTests/ModulesGraphTests.swift
@@ -409,7 +409,73 @@ final class ModulesGraphTests: XCTestCase {
             result.check(modules: "Bar", "Baz", "Foo")
         }
     }
-
+    
+    func testLibraryInvalidDependencyOnTestTarget() throws {
+        
+        let fs = InMemoryFileSystem(
+            emptyFiles:
+            "/Foo/Sources/Foo/Foo.swift",
+            "/Foo/Tests/FooTest/FooTest.swift"
+        )
+        
+        let observability = ObservabilitySystem.makeForTesting()
+        
+        let _ = try loadModulesGraph(
+            fileSystem: fs,
+            manifests: [
+                Manifest.createRootManifest(
+                    displayName: "Foo",
+                    path: "/Foo",
+                    toolsVersion: .v6_0,
+                    products: [
+                        ProductDescription(name: "Foo", type: .library(.automatic), targets: ["FooTest"]),
+                    ],
+                    targets: [
+                        TargetDescription(name: "Foo", dependencies: ["FooTest"]),
+                        TargetDescription(name: "FooTest", type: .test)
+                    ]
+                ),
+            ],
+            observabilityScope: observability.topScope)
+        
+        testDiagnostics(observability.diagnostics) { result in
+            result.check(
+                diagnostic: "Invalid dependency: 'Foo' cannot depend on test target dependency 'FooTest'. Only test targets can depend on other test targets",
+                severity: .error
+            )
+        }
+    }
+    
+    func testValidDependencyOnTestTarget() throws {
+        
+        let fs = InMemoryFileSystem(
+            emptyFiles:
+            "/Foo/Tests/Foo/Foo.swift",
+            "/Foo/Tests/FooTest/FooTest.swift"
+        )
+        
+        let observability = ObservabilitySystem.makeForTesting()
+        
+        let _ = try loadModulesGraph(
+            fileSystem: fs,
+            manifests: [
+                Manifest.createRootManifest(
+                    displayName: "Foo",
+                    path: "/Foo",
+                    toolsVersion: .v6_0,
+                    products: [
+                    ],
+                    targets: [
+                        TargetDescription(name: "Foo", dependencies: ["FooTest"], type: .test),
+                        TargetDescription(name: "FooTest", type: .test)
+                    ]
+                ),
+            ],
+            observabilityScope: observability.topScope)
+        
+        XCTAssertNoDiagnostics(observability.diagnostics)
+    }
+    
     // Make sure there is no error when we reference Test targets in a package and then
     // use it as a dependency to another package. SR-2353
     func testTestTargetDeclInExternalPackage() throws {

--- a/Tests/PackageGraphTests/ModulesGraphTests.swift
+++ b/Tests/PackageGraphTests/ModulesGraphTests.swift
@@ -446,6 +446,115 @@ final class ModulesGraphTests: XCTestCase {
         }
     }
 
+    func testExecutableInvalidDependencyOnTestTarget() throws {
+        let fs = InMemoryFileSystem(
+            emptyFiles:
+            "/Foo/Sources/Foo/main.swift",
+            "/Foo/Tests/FooTest/FooTest.swift"
+        )
+
+        let observability = ObservabilitySystem.makeForTesting()
+
+        let _ = try loadModulesGraph(
+            fileSystem: fs,
+            manifests: [
+                Manifest.createRootManifest(
+                    displayName: "Foo",
+                    path: "/Foo",
+                    toolsVersion: .v6_0,
+                    targets: [
+                        TargetDescription(name: "Foo", dependencies: ["FooTest"], type: .executable),
+                        TargetDescription(name: "FooTest", type: .test),
+                    ]
+                ),
+            ],
+            observabilityScope: observability.topScope
+        )
+
+        testDiagnostics(observability.diagnostics) { result in
+            result.check(
+                diagnostic: "Invalid dependency: 'Foo' cannot depend on test target dependency 'FooTest'. Only test targets can depend on other test targets",
+                severity: .error
+            )
+        }
+    }
+
+    func testPluginInvalidDependencyOnTestTarget() throws {
+        let fs = InMemoryFileSystem(
+            emptyFiles:
+            "/Foo/Plugins/Foo/main.swift",
+            "/Foo/Tests/FooTest/FooTest.swift"
+        )
+
+        let observability = ObservabilitySystem.makeForTesting()
+
+        let _ = try loadModulesGraph(
+            fileSystem: fs,
+            manifests: [
+                Manifest.createRootManifest(
+                    displayName: "Foo",
+                    path: "/Foo",
+                    toolsVersion: .v6_0,
+                    targets: [
+                        TargetDescription(
+                            name: "Foo",
+                            dependencies: ["FooTest"],
+                            type: .plugin,
+                            pluginCapability: .buildTool
+                        ),
+                        TargetDescription(name: "FooTest", type: .test),
+                    ]
+                ),
+            ],
+            observabilityScope: observability.topScope
+        )
+
+        testDiagnostics(observability.diagnostics) { result in
+            result.check(
+                diagnostic: "Invalid dependency: 'Foo' cannot depend on test target dependency 'FooTest'. Only test targets can depend on other test targets",
+                severity: .error
+            )
+        }
+    }
+    
+    func testMacroInvalidDependencyOnTestTarget() throws {
+        let fs = InMemoryFileSystem(
+            emptyFiles:
+            "/Foo/Sources/Foo/main.swift",
+            "/Foo/Tests/FooTest/FooTest.swift"
+        )
+
+        let observability = ObservabilitySystem.makeForTesting()
+
+        let _ = try loadModulesGraph(
+            fileSystem: fs,
+            manifests: [
+                Manifest.createRootManifest(
+                    displayName: "Foo",
+                    path: "/Foo",
+                    toolsVersion: .v6_0,
+                    targets: [
+                        TargetDescription(
+                            name: "Foo",
+                            dependencies: ["FooTest"],
+                            type: .macro
+                        ),
+                        TargetDescription(name: "FooTest", type: .test),
+                    ]
+                ),
+            ],
+            observabilityScope: observability.topScope
+        )
+
+        testDiagnostics(observability.diagnostics) { result in
+            result.check(
+                diagnostic: "Invalid dependency: 'Foo' cannot depend on test target dependency 'FooTest'. Only test targets can depend on other test targets",
+                severity: .error
+            )
+        }
+    }
+
+
     func testValidDependencyOnTestTarget() throws {
         let fs = InMemoryFileSystem(
             emptyFiles:

--- a/Tests/PackageGraphTests/ModulesGraphTests.swift
+++ b/Tests/PackageGraphTests/ModulesGraphTests.swift
@@ -409,17 +409,16 @@ final class ModulesGraphTests: XCTestCase {
             result.check(modules: "Bar", "Baz", "Foo")
         }
     }
-    
+
     func testLibraryInvalidDependencyOnTestTarget() throws {
-        
         let fs = InMemoryFileSystem(
             emptyFiles:
             "/Foo/Sources/Foo/Foo.swift",
             "/Foo/Tests/FooTest/FooTest.swift"
         )
-        
+
         let observability = ObservabilitySystem.makeForTesting()
-        
+
         let _ = try loadModulesGraph(
             fileSystem: fs,
             manifests: [
@@ -432,12 +431,13 @@ final class ModulesGraphTests: XCTestCase {
                     ],
                     targets: [
                         TargetDescription(name: "Foo", dependencies: ["FooTest"]),
-                        TargetDescription(name: "FooTest", type: .test)
+                        TargetDescription(name: "FooTest", type: .test),
                     ]
                 ),
             ],
-            observabilityScope: observability.topScope)
-        
+            observabilityScope: observability.topScope
+        )
+
         testDiagnostics(observability.diagnostics) { result in
             result.check(
                 diagnostic: "Invalid dependency: 'Foo' cannot depend on test target dependency 'FooTest'. Only test targets can depend on other test targets",
@@ -445,17 +445,16 @@ final class ModulesGraphTests: XCTestCase {
             )
         }
     }
-    
+
     func testValidDependencyOnTestTarget() throws {
-        
         let fs = InMemoryFileSystem(
             emptyFiles:
             "/Foo/Tests/Foo/Foo.swift",
             "/Foo/Tests/FooTest/FooTest.swift"
         )
-        
+
         let observability = ObservabilitySystem.makeForTesting()
-        
+
         let _ = try loadModulesGraph(
             fileSystem: fs,
             manifests: [
@@ -467,15 +466,16 @@ final class ModulesGraphTests: XCTestCase {
                     ],
                     targets: [
                         TargetDescription(name: "Foo", dependencies: ["FooTest"], type: .test),
-                        TargetDescription(name: "FooTest", type: .test)
+                        TargetDescription(name: "FooTest", type: .test),
                     ]
                 ),
             ],
-            observabilityScope: observability.topScope)
-        
+            observabilityScope: observability.topScope
+        )
+
         XCTAssertNoDiagnostics(observability.diagnostics)
     }
-    
+
     // Make sure there is no error when we reference Test targets in a package and then
     // use it as a dependency to another package. SR-2353
     func testTestTargetDeclInExternalPackage() throws {

